Show help messages of --dump

sonots opened the pull request:
https://bugs.ruby-lang.org/issues/13849

yui-knk commented on ruby.c:
```diff
@@ -271,6 +274,27 @@ usage(const char *name, int help)
+	unsigned short namelen, secondlen;
+    };
+    static const struct message usage_msg[] = {
+	M("--dump=insns",     "", "dump YARV instructions"),
```
Nice patch 🎉 

I feel "dump instruction sequences" is better. Because the name "YARV" is not exposed to public.

sonots replied:
okay

sonots commented:
> I've removed old --dump=help and --dump=usage, and changed --dump=help to show help messages of --dump.

@nobu could you check this change is okay?

sonots commented:
i want to get code reviews

knu commented on ruby.c:
```diff
@@ -241,6 +254,7 @@ usage(const char *name, int help)
     };
     static const struct message help_msg[] = {
 	M("--copyright",                   "", "print the copyright"),
+	M("--dump=argument[,...]",	   "", "argument is such as insns. see --help=dump for more details"),
```
You could list some of the valid arguments like so:
`M("--dump={insns|parsetree|...}[,...]", "", "dump debug information; see ..."),`

sonots replied:
thanks, fixed so

sonots replied:
Also, changed

```
  --enable=feature[,...], --disable=feature[,...]
                  enable or disable features
...
Features:
  gems            rubygems (default: enabled)
  did_you_mean    did_you_mean (default: enabled)
  rubyopt         RUBYOPT environment variable (default: enabled)
  frozen-string-literal
                  freeze all string literals (default: disabled)
```

to

```
  --enable={gems|did_you_mean|...}[,...], --disable={gems|did_you_mean|...}[,...]
                  enable or disable features shown below
...

...
Features:
  gems            rubygems (default: enabled)
  did_you_mean    did_you_mean (default: enabled)
  rubyopt         RUBYOPT environment variable (default: enabled)
  frozen-string-literal
                  freeze all string literals (default: disabled)
```

knu commented on ruby.c:
```diff
@@ -271,6 +284,23 @@ usage(const char *name, int help)
+          "dump AST or AST with comments"),
+    };
+    int i;
+    int help = 1; // used in SHOW
```
You should stick with C90 style comments in our source tree: `/* ... */`

sonots replied:
fixed

knu commented on ruby.c:
```diff
@@ -90,6 +90,16 @@ enum feature_flag_bits {
 
+#define HELP_BIT(bit) (1U << help_##bit)
+#define DEFINE_HELP(bit) help_##bit
+#define EACH_HELPS(X, SEP) \
```
Hmm, each <plural> doesn't sound right as English.  Should it be fixed along with `EACH_DUMPS`?

sonots replied:
I wonder which `EACH_HELP` or `HELPS_EACH` is better.
`HELPS_EACH` would be good as it looks like `helps.each {}`?

@nobu what do you think? `EACH_HELP` or `HELPS_EACH` or just keep?

knu commented:
This would be a great step forward for potential developers! 👍

sonots commented:
Based on a discussion at https://bugs.ruby-lang.org/issues/13849#note-17, I modified help message as:

```
$ ./ruby --help
Usage: ./ruby [switches] [--] [programfile] [arguments]
  -0[octal]       specify record separator (\0, if no argument)
  -a              autosplit mode with -n or -p (splits $_ into $F)
  -c              check syntax only
  -Cdirectory     cd to directory before executing your script
  -d, --debug     set debugging flags (set $DEBUG to true)
  -e 'command'    one line of script. Several -e's allowed. Omit [programfile]
  -Eex[:in], --encoding=ex[:in]
                  specify the default external and internal character encodings
  -Fpattern       split() pattern for autosplit (-a)
  -i[extension]   edit ARGV files in place (make backup if extension supplied)
  -Idirectory     specify $LOAD_PATH directory (may be used more than once)
  -l              enable line ending processing
  -n              assume 'while gets(); ... end' loop around your script
  -p              assume loop like -n but print line also like sed
  -rlibrary       require the library before executing your script
  -s              enable some switch parsing for switches after script name
  -S              look for the script using PATH environment variable
  -T[level=1]     turn on tainting checks
  -v, --verbose   print version number, then turn on verbose mode
  -w              turn warnings on for your script
  -W[level=2]     set warning level; 0=silence, 1=medium, 2=verbose
  -x[directory]   strip off text before #!ruby line and perhaps cd to directory
  --copyright     print the copyright
  --dump={insns|parsetree|...}[,...]
                  dump debug information. see below for available dump list
  --enable={gems|did_you_mean|...}[,...], --disable={gems|did_you_mean|...}[,...]
                  enable or disable features. see below for available features list
  --external-encoding=encoding, --internal-encoding=encoding
                  specify the default external or internal character encoding
  --version       print the version
  --help          show this message, -h for short message
Dump List:
  insns           instruction sequences
  yydebug         yydebug of yacc parser generator
  parsetree       AST
  parsetree_with_comment
                  AST with comments
Features List:
  gems            rubygems (default: enabled)
  did_you_mean    did_you_mean (default: enabled)
  rubyopt         RUBYOPT environment variable (default: enabled)
  frozen-string-literal
                  freeze all string literals (default: disabled)
```

sonots commented:
fixed via https://github.com/ruby/ruby/commit/c18503c25ab1d765603bee36d7683ef7a6085c65
